src/response/file.rs

```diff
@@ -0,0 +1,92 @@
+use http::header::{CONTENT_LENGTH, CONTENT_TYPE, ETAG, LAST_MODIFIED};
+use httpdate::HttpDate;
+use std::fs::Metadata;
+use std::io::{self, ErrorKind};
+use std::path::{Path, PathBuf};
+use tokio::fs;
+
+use super::Response;
+use crate::{middleware, Error};
+
+type GenerateEtag = fn(&Metadata) -> Result<String, Error>;
+
+/// Serve a single file from disk.
+///
+pub struct File {
+    path: PathBuf,
+    etag: Option<GenerateEtag>,
+    mime_type: Option<String>,
+    with_last_modified: bool,
+}
+
+fn wrap_io_error(error: io::Error) -> Error {
+    let kind = error.kind();
+    let source = Box::new(error);
+
+    match kind {
+        ErrorKind::PermissionDenied => Error::forbidden(source),
+        ErrorKind::NotFound => Error::not_found(source),
+        _ => Error::internal_server_error(source),
+    }
+}
+
+impl File {
+    pub fn open(path: impl AsRef<Path>) -> Self {
+        Self {
+            path: path.as_ref().to_path_buf(),
+            etag: None,
+            mime_type: None,
+            with_last_modified: false,
+        }
+    }
+
+    pub fn etag(self, f: GenerateEtag) -> Self {
+        Self {
+            etag: Some(f),
+            ..self
+        }
+    }
+
+    pub fn mime_type(self, mime_type: &str) -> Self {
+        Self {
+            mime_type: Some(mime_type.to_owned()),
+            ..self
+        }
+    }
+
+    pub fn with_last_modified(self) -> Self {
+        Self {
+            with_last_modified: true,
+            ..self
+        }
+    }
+
+    pub async fn serve(self) -> middleware::Result {
+        let path = self.path.as_path();
+        let data = fs::read(path).await.map_err(wrap_io_error)?;
+        let meta = if self.etag.is_some() || self.with_last_modified {
+            Some(fs::metadata(path).await.map_err(wrap_io_error)?)
+        } else {
+            None
+        };
+
+        let mut response = Response::build().header(CONTENT_LENGTH, data.len());
```

Comment:
I don't mind reassigning to `response` for the conditional parts of the builder. This code is sync for the remainder of the function.